keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,262 @@
+## Motivation
+
+In the current Endpoints API, one object instance contains all the individual endpoints of a service. Whenever a single pod in a service is added/updated/deleted, the whole Endpoints object (even when the other endpoints didn't change) is re-computed, written to storage (etcd) and sent to all watchers (e.g. kube-proxy). This leads to 2 major problems:
+
+- Storing multiple megabytes of endpoints puts strain on multiple parts of the system due to not having a paging system and a monolithic watch/storage design. [The max number of endpoints is bounded by the K8s storage layer (etcd)][max-object-size]], which has a hard limit on the size of a single object (1.5MB by default). That means attempts to write an object larger than the limit will be rejected. Additionally, there is a similar limitation in the watch path in Kubernetes apiserver. For a K8s service, if its Endpoints object is too large, endpoint updates will not be propagated to kube-proxy(s), and thus iptables/ipvs won’t be reprogrammed.
+- [Performance degradation in large k8s deployments.][perf-degrade] Not being able to efficiently read/update individual endpoint changes can lead to (e.g during rolling upgrade of a service) endpoints operations that are quadratic in the number of its elements. If one consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the quadratic traffic gets multiplied further with number of watches (usually equal to #nodes in the cluster).
+
+The new EndpointSlice API aims to address existing problems as well as leaving room for future extension.
+
+
+### Goal
+
+- Support tens of thousands of backend endpoints in a single service on cluster with thousands of nodes.
+- Leave room for foreseeable extension:
+  - Support multiple IPs per pod
+  - More endpoint states than Ready/NotReady
+  - Dynamic endpoint subsetting
+  
+### Non-Goal
+- Change functionality provided by K8s V1 Service API.
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the networking API group. 
+
+```
+type EndpointSlice struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec EndpointSliceSpec
+}
+
+type EndpointSliceSpec struct {
+        Endpoints []Endpoint
+        Ports []EndpointPort
+}
+
+type EndpointPort struct {
+        // Optional
+        Name string 
+        // Required
+        Protocol Protocol
+        // Optional: If unspecified, port remapping is not implemented
+        Port *int32
+}
+
+type Endpoint struct {
+        // Required: must contain at least one IP.
+        IPs []net.IP      
+        // Optional
+        Hostname string         
+        // Optional
+        NodeName *string  
+        // Optional
+        Condition EndpointCondition
+        // Optional
+        TargetRef *ObjectReference  
+}
+
+type EndpointCondition struct {
+       // Matches the Ready condition on pod
```

Comment:
Good point. 
If we want to completely decouple pod and endpoint status. Probably change it to `ReadyForTraffic`?

Reply:
I don't think "ready" is confusing, but we should document it all the same.  This also means we should think hard about what "containers ready" really means in the abstract space.  What will these 2 be used for?  How do we explain which field to use when?

Reply:
Probably missing some context, but this sounds more like a status than a spec?

Reply:
@apelisse This is per endpoint condition. I feel like status is mostly for the overall `EndpointSlice` object, not per individual endpoint. Also, EndpointsSlice controller generates the spec. The consumer of EndpointSlice objects should be responsible for populating status.

Reply:
@thockin 
I agree that `ContainersReady` should not be here. 

One feedback I got for Pod ReadinessGate is that it could not determine whether the readinessGate influences endpoint readiness. So the current states looks like:
 ``` Endpiont is Ready == Pod is Ready == Containers are ready AND Readiness Gates are True```

Maybe we should further evolve the states to something like this:
```
Endpiont is Ready == Containers are ready AND Readiness Gates that influence endpoint readiness are True```
Pod is Ready == Containers are ready AND all Readiness Gates are True`
```

Workloads consumes Pod readiness while services consumes endpoint readiness. This way we are able to express more states in finer grain so that we can implement better handshakes between workload and service. 

`Ready` sort of implied it correspond to Pod readiness. So that was why I was suggesting changing it to something different like `ReadyForTraffic`.

Reply:
I still believe this is an "Observed" state, rather than a "Desired" state and so it makes more sense in status than in spec. It's also "transient" as in "it changes even though the desired state doesn't change". 

I do understand that these are per EndPoint rather than for the slice (that's another drawback to the multiple things per object design). You could still have a list in status with the conditions, and maybe an aggregated condition if that makes sense.

Also in general, we avoid "booleans" because they don't work very well for extensibility and tend to break controllers state-machines. Lists are recommended.

Reply:
Let's focus on use-cases.

When an endpoint is not ready, a consumer of Endpoints can't do much with it right?

When an endpoint is X, it claims that it is ready for traffic (e.g a pod has all containers ready) but interlocked systems may not be synchronized (e.g. a pod has unresolved readiness gates).

When an endpoint is ready, all interlocked systems have checked in.

I'm trying to keep it arms-length from being defined in terms of other kubernetes constructs, but showing how we would map them.  Concretely, we would want to indicate X (to be named) when all containers are ready but before readiness gates.  One of the readiness gates is configuring the LBs from Endpoints, excluding not-ready ones but including X ones, right?

Is there a risk that some readiness gates are more impactful than others?  e.g. what if there was a firewall applier that failed open...I guess that would just be terrible..

Reply:
IF we are going to have `spec` and `status` then I *think* I agree that status makes sense for these.  Ostensibly the Endpoints controller would periodically resync these if they were lost.

If we ditch `spec` and keep EndpointSlice dumb, then it becomes a non-issue.

Reply:
Would you get something for free if you have the status sub-resource? Separate permission to write these, etc?